From: chasey-dev <ellenyoung0912@gmail.com>
Subject: [PATCH] nl80211: Pass FT key holder parameters to MTK driver

mt_wifi7 needs the NAS identifier, R1 key holder, AP address and
reassociation deadline to process FT management frames. SET_FT carries
these values in ap_11r_params.

nl80211 sends the parameters when SET_FT is advertised and propagates
command failures to AP setup. FtSupport must already be enabled in DAT:
mt_wifi7 otherwise returns success without installing the parameters.

Signed-off-by: chasey-dev <ellenyoung0912@gmail.com>
---
--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -2626,6 +2626,26 @@ int ieee802_11_build_ap_params(struct ho
 	params->multicast_to_unicast = hapd->conf->multicast_to_unicast;
 	params->pbss = hapd->conf->pbss;
 
+#if defined(CONFIG_DRIVER_NL80211_MTK) && defined(CONFIG_IEEE80211R_AP)
+	if (wpa_key_mgmt_ft(hapd->conf->wpa_key_mgmt) &&
+	    hapd->conf->nas_identifier) {
+		params->mtk_ft_nas_identifier_len =
+			os_strlen(hapd->conf->nas_identifier);
+		if (params->mtk_ft_nas_identifier_len <=
+		    sizeof(params->mtk_ft_nas_identifier)) {
+			os_memcpy(params->mtk_ft_nas_identifier,
+				  hapd->conf->nas_identifier,
+				  params->mtk_ft_nas_identifier_len);
+			os_memcpy(params->mtk_ft_r1_key_holder,
+				  hapd->conf->r1_key_holder, ETH_ALEN);
+			os_memcpy(params->mtk_ft_own_mac, hapd->own_addr,
+				  ETH_ALEN);
+			params->mtk_ft_reassociation_deadline =
+				hapd->conf->reassociation_deadline;
+		}
+	}
+#endif /* CONFIG_DRIVER_NL80211_MTK && CONFIG_IEEE80211R_AP */
+
 	if (hapd->conf->ftm_responder) {
 		if (hapd->iface->drv_flags & WPA_DRIVER_FLAGS_FTM_RESPONDER) {
 			params->ftm_responder = 1;
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -1756,6 +1756,17 @@ struct wpa_driver_ap_params {
 	 */
 	int pbss;
 
+#if defined(CONFIG_DRIVER_NL80211_MTK) && defined(CONFIG_IEEE80211R_AP)
+	/**
+	 * mtk_ft_* - MTK driver FT parameters for AP-side SME 802.11r IEs
+	 */
+	u8 mtk_ft_nas_identifier[64];
+	size_t mtk_ft_nas_identifier_len;
+	u8 mtk_ft_r1_key_holder[ETH_ALEN];
+	u8 mtk_ft_own_mac[ETH_ALEN];
+	u32 mtk_ft_reassociation_deadline;
+#endif /* CONFIG_DRIVER_NL80211_MTK && CONFIG_IEEE80211R_AP */
+
 	/**
 	 * multicast_to_unicast - Whether to use multicast_to_unicast
 	 *
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -45,6 +45,11 @@
 #include "rfkill.h"
 #include "driver_nl80211.h"
 
+#if defined(CONFIG_DRIVER_NL80211_MTK) && defined(CONFIG_IEEE80211R_AP)
+static int nl80211_mtk_set_ft_params(struct i802_bss *bss,
+				struct wpa_driver_ap_params *params);
+#endif /* CONFIG_DRIVER_NL80211_MTK && CONFIG_IEEE80211R_AP */
+
 
 #ifndef NETLINK_CAP_ACK
 #define NETLINK_CAP_ACK 10
@@ -5295,6 +5300,67 @@ static int nl80211_put_freq_params(struc
 }
 
 
+#if defined(CONFIG_DRIVER_NL80211_MTK) && defined(CONFIG_IEEE80211R_AP)
+static int nl80211_mtk_set_ft_params(struct i802_bss *bss,
+				struct wpa_driver_ap_params *params)
+{
+	struct ap_11r_params ft_params;
+	struct nl_msg *msg;
+	struct nlattr *data;
+	int ret;
+
+	if (!bss->drv->mtk_set_ft || !params->mtk_ft_nas_identifier_len)
+		return 0;
+
+	if (params->mtk_ft_nas_identifier_len >
+	    sizeof(ft_params.nas_identifier))
+		return -EINVAL;
+
+	os_memset(&ft_params, 0, sizeof(ft_params));
+	ft_params.nas_id_len = params->mtk_ft_nas_identifier_len;
+	os_memcpy(ft_params.nas_identifier, params->mtk_ft_nas_identifier,
+		  params->mtk_ft_nas_identifier_len);
+	os_memcpy(ft_params.r1_key_holder, params->mtk_ft_r1_key_holder,
+		  ETH_ALEN);
+	os_memcpy(ft_params.own_mac, params->mtk_ft_own_mac, ETH_ALEN);
+	ft_params.reassociation_deadline =
+		params->mtk_ft_reassociation_deadline;
+
+	wpa_printf(MSG_DEBUG,
+		   "nl80211: MTK set FT parameters ifname=%s nas_id_len=%d r1kh="
+		   MACSTR " own=" MACSTR " reassoc_deadline=%u",
+		   bss->ifname, ft_params.nas_id_len,
+		   MAC2STR(ft_params.r1_key_holder),
+		   MAC2STR(ft_params.own_mac),
+		   ft_params.reassociation_deadline);
+
+	msg = nl80211_bss_msg(bss, 0, NL80211_CMD_VENDOR);
+	if (!msg ||
+	    nla_put_u32(msg, NL80211_ATTR_VENDOR_ID, MTK_NL80211_VENDOR_ID) ||
+	    nla_put_u32(msg, NL80211_ATTR_VENDOR_SUBCMD,
+			MTK_NL80211_VENDOR_SUBCMD_SET_FT) ||
+	    !(data = nla_nest_start(msg, NL80211_ATTR_VENDOR_DATA)) ||
+	    nla_put(msg, MTK_NL80211_VENDOR_ATTR_SET_FTIE,
+		    sizeof(ft_params), &ft_params))
+		goto fail;
+
+	nla_nest_end(msg, data);
+
+	ret = send_and_recv_cmd(bss->drv, msg);
+	if (ret)
+		wpa_printf(MSG_ERROR,
+			   "nl80211: MTK failed to set FT parameters ifname=%s ret=%d (%s)",
+			   bss->ifname, ret, strerror(-ret));
+
+	return ret;
+
+fail:
+	nlmsg_free(msg);
+	return -ENOBUFS;
+}
+#endif /* CONFIG_DRIVER_NL80211_MTK && CONFIG_IEEE80211R_AP */
+
+
 static int wpa_driver_nl80211_set_ap(void *priv,
 				     struct wpa_driver_ap_params *params)
 {
@@ -5327,6 +5393,12 @@ static int wpa_driver_nl80211_set_ap(voi
 		return -EINVAL;
 	}
 
+#if defined(CONFIG_DRIVER_NL80211_MTK) && defined(CONFIG_IEEE80211R_AP)
+	ret = nl80211_mtk_set_ft_params(bss, params);
+	if (ret)
+		return ret;
+#endif /* CONFIG_DRIVER_NL80211_MTK && CONFIG_IEEE80211R_AP */
+
 	beacon_set = params->reenable ? 0 : link->beacon_set;
 
 	wpa_printf(MSG_DEBUG, "nl80211: Set beacon (beacon_set=%d)",
--- a/src/drivers/driver_nl80211.h
+++ b/src/drivers/driver_nl80211.h
@@ -196,6 +196,7 @@ struct wpa_driver_nl80211_data {
 	unsigned int qca_do_acs:1;
 	unsigned int brcm_do_acs:1;
 	unsigned int mtk_do_acs:1;
+	unsigned int mtk_set_ft:1;
 	unsigned int uses_6ghz:1;
 	unsigned int uses_s1g:1;
 	unsigned int secure_ranging_ctx_vendor_cmd_avail:1;
--- a/src/drivers/driver_nl80211_capa.c
+++ b/src/drivers/driver_nl80211_capa.c
@@ -1159,6 +1159,9 @@ static int wiphy_info_handler(struct nl_
 						WPA_DRIVER_FLAGS_SUPPORT_HW_MODE_ANY;
 					drv->mtk_do_acs = 1;
 					break;
+				case MTK_NL80211_VENDOR_SUBCMD_SET_FT:
+					drv->mtk_set_ft = 1;
+					break;
 				}
 #endif /* CONFIG_DRIVER_NL80211_MTK */
 			}
